src/tui/mod.rs

```diff
@@ -121,27 +123,70 @@ fn render(tui: &Tui, area: Rect, buffer: &mut Buffer) -> miette::Result<()> {
 
 const SCROLL_AMOUNT: usize = 1;
 
-#[instrument(level = "trace", skip(tui))]
-fn handle_event(tui: &mut Tui, event: Event) -> miette::Result<()> {
+#[instrument(level = "trace", skip(tui, frame))]
+fn handle_event(tui: &mut Tui, event: Event, frame: Frame<'_>) -> miette::Result<()> {
+    let last_line = tui
+        .scrollback
+        .split(|byte| *byte == b'\n')
+        .count()
+        .saturating_sub(1);
+
     match event {
         Event::Mouse(mouse) if mouse.kind == MouseEventKind::ScrollUp => {
             tui.scroll_offset = tui.scroll_offset.saturating_sub(SCROLL_AMOUNT);
         }
         Event::Mouse(mouse) if mouse.kind == MouseEventKind::ScrollDown => {
-            let last_line = tui
-                .scrollback
-                .split(|byte| *byte == b'\n')
-                .count()
-                .saturating_sub(1);
-            tui.scroll_offset = min(last_line, tui.scroll_offset + SCROLL_AMOUNT);
-        }
-        Event::Key(key)
-            if key.modifiers == KeyModifiers::CONTROL && key.code == KeyCode::Char('c') =>
-        {
-            tui.quit = true;
+            tui.scroll_offset += SCROLL_AMOUNT;
         }
+        Event::Key(key) => match key.modifiers {
+            KeyModifiers::NONE => match key.code {
+                KeyCode::Char('j') => {
+                    tui.scroll_offset += 1;
+                }
+                KeyCode::Char('k') => {
+                    tui.scroll_offset = tui.scroll_offset.saturating_sub(1);
+                }
+                KeyCode::Char('g') => {
+                    tui.scroll_offset = 0;
+                }
+                _ => {}
+            },
+
+            #[allow(clippy::single_match)]
+            KeyModifiers::SHIFT => match key.code {
+                KeyCode::Char('G') => {
+                    tui.scroll_offset = last_line;
+                }
+                _ => {}
+            },
```

Comment:
You might be interested in the declarative macros I wrote for my text editor to match on `crossterm`'s `Key`s more concisely:

- [Usage example](https://github.com/evanrelf/indigo/blob/7a5e8e47291585cae03cdf5a7c47ad3bcd8db3e6/crates/indigo-tui/src/main.rs#L283-L307)
  ```rust
    match event {
        Event::Key(key) if k!(key, Left) || k!(key, CONTROL, 'b') => {
            command_mode.move_backward(1);
        }
        Event::Key(key) if k!(key, Right) || k!(key, CONTROL, 'f') => {
            command_mode.move_forward(1);
        }
        Event::Key(key) if k!(key, Left) || k!(key, ALT, 'b') => {
            command_mode.move_backward_word();
        }
    // ...
  ```
- [Implementation](https://github.com/evanrelf/indigo/blob/7a5e8e47291585cae03cdf5a7c47ad3bcd8db3e6/crates/indigo-tui/src/key/macros.rs)